### PR TITLE
Set Callbacks in Fuzzer and Add CHECKS in App

### PR DESCRIPTION
**Pull request by florian-kuebler**
Recent changes from "if (..._callback_)" to "CHECK(..._callback_)"
led to failed checks in the Fuzzers, as the callbacks were not
explicitly set.

This change sets the required callbacks in the Fuzzer.
Also it replaces the remaining "if"s with "CHECK" to ensure
that the callbacks are properly initialized.

Bug: http://b/168770539
Test: Compile

### CODE REVIEW

**Comment by dpallotti** [comment]
As Ronald wrote in the bug, this renders the `CHECK`s useless. Either we accept it and remove the `CHECK`s, or the initialization to dummy functions should happen in the fuzzer(s).

**Comment by florian-kuebler** [comment]
As written in the bug. These checks are basically explicitly stating the assumption that the function is save to call. Yes, currently this will be the case, but they should prevent us from changes in the future.

However, if you both strongly argue to remove the checks, that would be also fine for me to remove them. I would however, strongly avoid setting the callbacks in the fuzzer.

**Comment by dpallotti** [comment]
> As written in the bug. These checks are basically explicitly stating the assumption that the function is save to call. Yes, currently this will be the case, but they should prevent us from changes in the future.
> 
> However, if you both strongly argue to remove the checks, that would be also fine for me to remove them. I would however, strongly avoid setting the callbacks in the fuzzer.

On my side, the point is not about removing the `CHECK`s, but rather about the fact that they no longer guarantee that clients of `OrbitApp` set the callbacks and whether we are fine with this or not. So the question is, why

> I would however, strongly avoid setting the callbacks in the fuzzer.

? Is it because they are many and it's messy to set them all for each fuzzer?

**Comment by florian-kuebler** [comment]
> > As written in the bug. These checks are basically explicitly stating the assumption that the function is save to call. Yes, currently this will be the case, but they should prevent us from changes in the future.
> > However, if you both strongly argue to remove the checks, that would be also fine for me to remove them. I would however, strongly avoid setting the callbacks in the fuzzer.
> 
> On my side, the point is not about removing the `CHECK`s, but rather about the fact that they no longer guarantee that clients of `OrbitApp` set the callbacks and whether we are fine with this or not. So the question is, why
> 
> > I would however, strongly avoid setting the callbacks in the fuzzer.
> 
> ? Is it because they are many and it's messy to set them all for each fuzzer?

I see your point. My point was basically, not messing up the fuzzer here, you are right. Still I have the feeling, that a controller (the app), should be in a well behaving state after just constructing it.

File: OrbitGl/App.cpp [hunk 1]
```diff
@@ -131,11 +131,11 @@ void OrbitApp::OnCaptureStarted(int32_t process_id, std::string process_name,
         capture_data_ = CaptureData(process_id, std::move(process_name), std::move(process),
```
Comment (dimitry-) [inline OrbitGl/App.cpp:131]: this is comment on the commit message: s/Dummy/Empty/

**Review by dimitry- — APPROVED** [review]
Otherwise lgtm

**Comment by dimitry-** [comment]
As for checks... someone could still set these to nullptr if I am not mistaken.. But I agree they look less useful now. Maybe move these CHECKs to Set*Listener functions?

**Comment by florian-kuebler** [comment]
> As for checks... someone could still set these to nullptr if I am not mistaken.. But I agree they look less useful now. Maybe move these CHECKs to Set*Listener functions?

Just as a note, the callbacks are all value type functions. It is just that the default constructor of std::function will create a function that fails when it is called.
In general, I like the idea of moving the checks to the setter.

**Comment by florian-kuebler** [comment]
Thanks for your reviews so far. I moved the CHECKS to the setter methods

**Comment by ronaldfw** [comment]
> As written in the bug. These checks are basically explicitly stating the assumption that the function is save to call. Yes, currently this will be the case, but they should prevent us from changes in the future.

You are right that the CHECKs are still useful, but they lose a lot of what they were guarding us against before, which is: At some point between creation of the object and calling these methods, these callbacks have been properly initialized. That logic is potentially fragile and I'm not in favor of losing the guards against that. (Echoing what @dpallotti dpallotti said, I think.)

Moving the CHECKs to the setter as suggested by @dimitry- helps a bit, but it doesn't guarantee that the Set* methods are actually ever called (which is what we have as a guarantee now). 

If possible, I'd prefer to fix the root cause in the fuzzer. Can you clarify why you don't like that approach?

**Comment by ronaldfw** [comment]
> I see your point. My point was basically, not messing up the fuzzer here, you are right. Still I have the feeling, that a controller (the app), should be in a well behaving state after just constructing it.

But is it even possible that the app is in a well behaving state just after constructing it? If the callbacks are default initialized to do nothing, then you can call them (in the places where we had the CHECKs), but the UI is broken (unless the Set* methods are properly called, which we would have no guarantee for anymore). I'd not consider that well behaving.

**Comment by florian-kuebler** [comment]
> > I see your point. My point was basically, not messing up the fuzzer here, you are right. Still I have the feeling, that a controller (the app), should be in a well behaving state after just constructing it.
> 
> But is it even possible that the app is in a well behaving state just after constructing it? If the callbacks are default initialized to do nothing, then you can call them (in the places where we had the CHECKs), but the UI is broken (unless the Set* methods are properly called, which we would have no guarantee for anymore). I'd not consider that well behaving.

I'd argue that this depends on the definition of well behaving. The whole idea of the callbacks is to decouple our controller (the app) from the UI. So a controller that has not called the Set* methods, can still count on the behaviours of functions, but with having no visual response. Which is especially true for the fuzzer, but might also be relevant for e.g. tests.

As I said, I also see your point. I would loop in @beckerhe, as this would basically affect the fuzzers.

**Comment by ronaldfw** [comment]
> I'd argue that this depends on the definition of well behaving. The whole idea of the callbacks is to decouple our controller (the app) from the UI. So a controller that has not called the Set* methods, can still count on the behaviours of functions, but with having no visual response. Which is especially true for the fuzzer, but might also be relevant for e.g. tests.

The case we care about the most is the behavior when the app is in front of a user. Maybe there is a different way to guarantee that callbacks have been properly initialized or maybe the fuzzer could use a different type of App class, etc. I don't have insight into all parts that are affected here, but I'd like use to carefully consider what we are giving up here.

**Comment by florian-kuebler** [comment]
> > I'd argue that this depends on the definition of well behaving. The whole idea of the callbacks is to decouple our controller (the app) from the UI. So a controller that has not called the Set* methods, can still count on the behaviours of functions, but with having no visual response. Which is especially true for the fuzzer, but might also be relevant for e.g. tests.
> 
> The case we care about the most is the behavior when the app is in front of a user. Maybe there is a different way to guarantee that callbacks have been properly initialized or maybe the fuzzer could use a different type of App class, etc. I don't have insight into all parts that are affected here, but I'd like use to carefully consider what we are giving up here.

Right. As I said, I have no strong opinion on this. So I changed the code accordingly (default initiazlization and checks in the App and Set* calls in the Fuzzer).

**Review by ronaldfw — APPROVED** [review]
Please update the PR description, o/w LGTM!

**Comment by florian-kuebler** [comment]
Done. Will merge now. Thanks for the review!.

**Pull request merged**